### PR TITLE
compact hash map

### PR DESCRIPTION
**Pull request by bogdad**
Hi!

finished open addressing hash map and wanted to try to use it in game,
swapped routing info routes from CDict to CHashMap. 
not sure if performance is better yet, it might not be,
was able to run the game - it crashed and after some commits it was kind of stable :)

please advise how do you think its best to use the map)

### CODE REVIEW

File: lib/compact/src/compact_array.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    pub fn with_capacity(cap: usize) -> CompactArray<T, A> {
+        let mut vec = CompactArray {
+            ptr: PointerToMaybeCompact::default(),
+            cap: cap,
+            _alloc: PhantomData,
+        };
+
+        vec.ptr.set_to_free(A::allocate::<T>(cap));
+        for i in 0..cap {
+            vec.ptr.initialize_with_default(i as isize);
```
Comment (bogdad) [inline lib/compact/src/compact_array.rs]: that asks to be vectorized somehow)

File: lib/compact/src/compact_array.rs [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            Some(item)
+        } else {
+            None
+        }
+    }
+}
+
+impl<T, A: Allocator> Drop for IntoIter<T, A> {
+    fn drop(&mut self) {
+        println!("dropping");
```
Comment (bogdad) [inline lib/compact/src/compact_array.rs]: will remove

File: lib/compact/src/compact_hash_map.rs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+                self.entries[index].used = false;
+                self.size -= 1;
+                return Some(self.entries[index].value.clone());
+            }
+        }
+        None
+    }
+
+    fn ensure_capacity(&mut self) {
+        if 10 * self.size > 7 * self.entries.capacity() {
```
Comment (bogdad) [inline lib/compact/src/compact_hash_map.rs]: might fail on large sizes

File: lib/compact/src/compact_vec.rs [hunk 4]
```diff
@@ -87,6 +91,18 @@ impl<T: Compact + Clone, A: Allocator> CompactVec<T, A> {
         }
     }
 
+    pub fn push_at(&mut self, i: usize, value: T) {
```
Comment (bogdad) [inline lib/compact/src/compact_vec.rs]: thats probably not needed, will remove

File: lib/kay/src/type_registry.rs [hunk 5]
```diff
@@ -16,6 +16,12 @@ impl ShortTypeId {
     }
 }
 
+impl Default for ShortTypeId {
+    fn default() -> Self {
```
Comment (bogdad) [inline lib/kay/src/type_registry.rs]: dont like it, but Default is very useful, not sure how to proceed

File: src/game/lanes_and_cars/pathfinding/mod.rs [hunk 6]
```diff
@@ -248,19 +248,20 @@ pub fn setup(system: &mut ActorSystem) {
                 })
                 .unwrap_or(true);
             if join {
+                let tell_to_forget_next_tick = lane.pathfinding
```
Comment (bogdad) [inline src/game/lanes_and_cars/pathfinding/mod.rs]: had to rearrange as it failed with borrowing errors I did not have time to dig into

**Comment by aeickhoff** [comment]
Hi again and thank you very much!

I will try out your version soon, I wasn't aware that you wanted to use your hash map for routing tables, sounds very promising - I hope the performance will indeed be better! I will profile it and let you know about the results here!

Also, I am currently working on some slightly conflicting changes to `compact`, I will (probably in a couple days) use the branch where I do that as the base for this PR and do the small changes that will be necessary.

**Comment by bogdad** [comment]
Hi!

Sure, no problem. I’ve just wanted to try if chashmap could be used in place of cdict, and see if it would work at all. It works, but I can’t see any difference in performance :) (https://github.com/bogdad/citybound/blob/perf/perf/flamegraph_compacthashmap.svg, https://github.com/bogdad/citybound/blob/perf/perf/flamegraph_master.svg) 

Also I noticed that rust’s default hash map does not need keys or values implementing Default, so I guess openadressingmap should not too, I’ll try to fix it in the future somehow.

**Comment by aeickhoff** [comment]
(I'll fix the conflicts, bring everything up to date and do performance tests)

**Comment by aeickhoff** [comment]
Continued in #192

**Pull request merged**